File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Components/Factories/VButtonFactory.kt
package com.vuenative.core

import android.content.Context
import android.view.View
import android.widget.TextView
import com.google.android.flexbox.FlexboxLayout

class VButtonFactory : NativeComponentFactory {
    override fun createView(context: Context): View {
        return TouchableView(context)
    }

    override fun updateProp(view: View, key: String, value: Any?) {
        val touchable = view as? TouchableView
        if (touchable == null) {
            StyleEngine.apply(key, value, view)
            return
        }

        when (key) {
            "disabled" -> {
                touchable.isDisabled = value == true || value == "true" ||
                        (value is Number && value.toInt() != 0)
            }
            "activeOpacity" -> {
                touchable.activeOpacity = when (value) {
                    is Double -> value.toFloat()
                    is Float -> value
                    is Int -> value.toFloat()
                    is String -> value.toFloatOrNull() ?: 0.7f
                    else -> 0.7f
                }
            }
            "title" -> {
                val flex = view as? FlexboxLayout
                val tv = (0 until (flex?.childCount ?: 0))
                    .mapNotNull { flex?.getChildAt(it) as? TextView }
                    .firstOrNull()
                tv?.text = value?.toString() ?: ""
            }
            else -> StyleEngine.apply(key, value, view)
        }
    }

    override fun addEventListener(view: View, event: String, handler: (Any?) -> Unit) {
        val touchable = view as? TouchableView ?: return

        when (event) {
            "press" -> touchable.onPress = { handler(null) }
            "longPress", "longpress" -> touchable.onLongPress = { handler(null) }
        }
    }

    override fun removeEventListener(view: View, event: String) {
        val touchable = view as? TouchableView ?: return

        when (event) {
            "press" -> touchable.onPress = null
            "longPress", "longpress" -> touchable.onLongPress = null
        }
    }

    override fun insertChild(parent: View, child: View, index: Int) {
        val flex = parent as? FlexboxLayout ?: return
        val lp = StyleEngine.buildFlexLayoutParams(child)
        if (index >= flex.childCount) flex.addView(child, lp)
        else flex.addView(child, index, lp)
    }

    override fun removeChild(parent: View, child: View) {
        (parent as? FlexboxLayout)?.removeView(child)
    }
}

File: native/ios/VueNativeCore/Sources/VueNativeCore/Components/Factories/VButtonFactory.swift
#if canImport(UIKit)
import UIKit
import ObjectiveC
import FlexLayout

/// Factory for VButton — the touchable/pressable component.
/// Maps to a TouchableView (custom UIView subclass) with FlexLayout enabled.
/// Supports press and long press events with configurable active opacity.
final class VButtonFactory: NativeComponentFactory {

    // MARK: - Associated object keys for event handlers

    private static var pressHandlerKey: UInt8 = 0
    private static var longPressHandlerKey: UInt8 = 0

    // MARK: - NativeComponentFactory

    func createView() -> UIView {
        let touchable = TouchableView()
        // Accessing .flex automatically enables Yoga layout
        _ = touchable.flex
        return touchable
    }

    func updateProp(view: UIView, key: String, value: Any?) {
        guard let touchable = view as? TouchableView else {
            // Fallback to StyleEngine for non-TouchableView
            StyleEngine.apply(key: key, value: value, to: view)
            return
        }

        switch key {
        case "disabled":
            if let disabled = value as? Bool {
                touchable.isDisabled = disabled
            } else if let disabled = value as? Int {
                touchable.isDisabled = disabled != 0
            } else {
                touchable.isDisabled = false
            }

        case "activeOpacity":
            if let opacity = value as? Double {
                touchable.activeOpacity = CGFloat(opacity)
            } else if let opacity = value as? Int {
                touchable.activeOpacity = CGFloat(opacity)
            } else {
                touchable.activeOpacity = 0.7
            }

        default:
            // Delegate to StyleEngine for layout/visual styling
            StyleEngine.apply(key: key, value: value, to: view)
        }
    }

    func addEventListener(view: UIView, event: String, handler: @escaping (Any?) -> Void) {
        guard let touchable = view as? TouchableView else { return }

        switch event {
        case "press":
            // Capture handler weakly via the closure to avoid retain cycles.
            // TouchableView.onPress holds the only strong reference to the closure.
            touchable.onPress = {
                handler(nil)
            }

        case "longpress":
            touchable.onLongPress = {
                handler(nil)
            }

        default:
            break
        }
    }

    func removeEventListener(view: UIView, event: String) {
        guard let touchable = view as? TouchableView else { return }

        switch event {
        case "press":
            touchable.onPress = nil

        case "longpress":
            touchable.onLongPress = nil

        default:
            break
        }
    }
}
#endif
